File: Kafka-Utils-Java/src/main/java/com/yhyr/comsumer/CommitSpecialOffset.java
package com.yhyr.comsumer;

import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.consumer.OffsetAndMetadata;
import org.apache.kafka.common.PartitionInfo;
import org.apache.kafka.common.TopicPartition;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;

/**
 * Class Subject: 修改给定Group、指定Topic下各Partition的offset
 * 
 * Tips: offset的修改必须是基于Partition而言的, 而非是Topic
 *
 * @author yhyr
 * @since 2018/12/30 18:25
 */
public class CommitSpecialOffset {
    private static Map<TopicPartition, OffsetAndMetadata> currentOffsets = new HashMap<>();

    public static void main(String[] args) {
        String brokers = "localhost:9092";
        String group = "group_test";
        String topic = "topic_demo";

        Properties props = new Properties();
        props.put("bootstrap.servers", brokers);
        props.put("group.id", group);
        props.put("enable.auto.commit", "true");
        props.put("auto.offset.reset", "latest");
        props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        props.put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props);

        // 获取Topic的Partition信息
        List<PartitionInfo> partitionInfos = consumer.partitionsFor(topic);
        // 将所有Partition的offset设置为10
        int resetOffsetValue = 10;
        partitionInfos.forEach(partitionInfo -> currentOffsets.put(new TopicPartition(partitionInfo.topic(),
            partitionInfo.partition()), new OffsetAndMetadata(resetOffsetValue)));
        consumer.commitSync(currentOffsets);

    }
}

File: Kafka-Utils-Python/consumer/commit_special_offset.py
# -*- coding: utf-8 -*-
"""
Subject: 修改给定Group、Topic下各Partition的offset; 不触发Group的Rebalance
    Tips: offset的修改必须是基于Partition而言的, 而非是Topic

@Author YH YR
@Time 2018/12/31 14:10
"""
from kafka import KafkaConsumer, TopicPartition, OffsetAndMetadata


class CommitSpecialOffset:
    def __init__(self, broker_list, group_name, topic):
        self.topic = topic
        self.consumer = KafkaConsumer(group_id=group_name, bootstrap_servers=broker_list)

    def reset_offset(self, reset_offset_value):
        partitions_offset = {}
        for partition_id in self.consumer.partitions_for_topic(self.topic):
            partitions_offset[TopicPartition(topic, partition_id)] = OffsetAndMetadata(reset_offset_value, '')

        self.consumer.commit(partitions_offset)


if __name__ == '__main__':
    broker_list = 'localhost:9092'
    group_name = 'group_test'
    topic = 'topic_demo'

    action = CommitSpecialOffset(broker_list, group_name, topic)
    action.reset_offset(500)
